ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -675,40 +609,93 @@ class ( HasHeader blk
-      (ExtLedgerState blk)
-      -- ^ The ledger state corresponding to the tip of the valid prefix of
-      -- the chain.
+type InvalidBlocks blk = Map (HeaderHash blk) (InvalidBlockReason blk, SlotNo)
+
+-- | Result of 'validate', also used internally.
+data ValidationResult blk =
+    ValidationResult
+      !(Chain blk)           -- ^ Valid prefix
+      !(ExtLedgerState blk)  -- ^ Corresponds to the tip of the valid prefix
+      !(InvalidBlocks blk)   -- ^ Invalid blocks encountered while validating
+                             -- the candidate chain.
 
+-- | Validate the given 'Chain'.
+--
+-- The 'InvalidBlocks' in the returned 'ValidationResult' will be >= the
+-- 'invalid' of the given 'Model'.
 validate :: forall blk. LedgerSupportsProtocol blk
          => TopLevelConfig blk
-         -> ExtLedgerState blk
+         -> Model blk
          -> Chain blk
          -> ValidationResult blk
-validate cfg initLedger chain =
-    go initLedger Genesis (Chain.toOldestFirst chain)
+validate cfg Model { currentSlot, maxClockSkew, initLedger, invalid } chain =
+    go initLedger Genesis (Chain.toOldestFirst chain) Nothing
   where
     go :: ExtLedgerState blk  -- ^ Corresponds to the tip of the valid prefix
        -> Chain blk           -- ^ Valid prefix
        -> [blk]               -- ^ Remaining blocks to validate
+       -> Maybe (ValidationResult blk)
+          -- ^ If 'Nothing', continue as normal. If 'Just', continue
+          -- validating, but return the chain and ledger from this
+          -- 'ValidationResult', augmented with any new invalid blocks.
+          --
+          -- This is used to continue validating blocks, even when they're
+          -- from the future, as the real implementation would also validate
+          -- them. This makes sure we detect the same invalid blocks in the
+          -- future.
        -> ValidationResult blk
-    go ledger validPrefix bs = case bs of
-      []    -> ValidChain validPrefix ledger
+    go ledger validPrefix bs mbFinal = case bs of
+      -- Return 'mbFinal' if it contains an "earlier" result
+      []    -> fromMaybe (ValidationResult validPrefix ledger invalid) mbFinal
       b:bs' -> case runExcept (tickThenApply cfg b ledger) of
-        Right ledger' -> go ledger' (validPrefix :> b) bs'
-        Left  e       -> InvalidChain e (fmap blockRealPoint (b NE.:| bs'))
-                           validPrefix ledger
+        -- Invalid block according to the ledger
+        Left e
+          -> ValidationResult validPrefix ledger (invalid <> invalid')
+          where
+            invalid' :: InvalidBlocks blk
+            invalid' =
+              Map.insert
+                (Block.blockHash b)
+                (ValidationError e, Block.blockSlot b) $
+              Map.fromList
+                [ (Block.blockHash b', (reason, Block.blockSlot b'))
+                | b' <- bs'
+                , let reason = InChainAfterInvalidBlock (blockRealPoint b) e
+                ]
+
```

Comment:
Add the following to ChainDB.md and update the implementation to not do the `InChainAfterInvalidBlock` thing.

CS Candidate: A -> I -> C

Why do we have to mark C as invalid?
Will there ever be a CS candidate containing C but not I?

My conjecture: no
Any chain containing C must also contain I
not every fragment must be anchored at the same point, *but* it must be anchored on our chain, and since I is invalid, we will never adopt I. So if a candidate fragment contains C and is anchored on our chain, it must also contain I.